### PR TITLE
chore(tests): create npm module from playwright tests

### PR DESCRIPTION
**Pull request by odockal**
### What does this PR do?
Allows `tests/playwright` folder to also compose an npm package and publish it.
Note: WIP: I would like to get some tests on this and opinion how to properly compose structure and the outcome at npmjs.com since I am not sure what exactly `yarn publish` will do.

### What issues does this PR fix or reference?
#6441 

### How to test this PR?
```
cd tests/playwright
yarn
npm pack
```
The result is tar archive (podman-desktop-tests-playwright) which can be imported as dependnecy in other projects. there is also readme explaining steps to follow.

- [ ] Tests are covering the bug fix or the new feature

### CODE REVIEW

File: tests/playwright/README.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+1. Node 20 (ideal is to use `nvm`)
+2. Install local dependencies: `npm install`
+3. Build: `npm run build`
+4. Create Local package `npm run package`
+5. Connect to the repository where you want to write your tests
+
+## Usage
+
+1. Add necessary dependencies, ie. devDependencies: "podman-desktop-tests-playwright": "0.0.1"
```
Comment (benoitf) [inline tests/playwright/README.md]:
```suggestion
1. Add necessary dependencies, ie. devDependencies: `"@podman-desktop/tests-playwright": "^1.8.0"`

```

File: tests/playwright/README.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+2. Additional dependencies like vitest or playwright
+
+### Test Runner Context and Hook extending
+
+Extending afterEach Hook using custom TestContext
+In your project, you need to define the TestContext interface to be passed into extended hook
+
+```runner-test-context.ts
+import type { TestContext } from 'vitest';
+import { PodmanDesktopRunner } from 'podman-desktop-tests-playwright';
```
Comment (benoitf) [inline tests/playwright/README.md]:
```suggestion
import { PodmanDesktopRunner } from '@podman-desktop/tests-playwright';
```

File: tests/playwright/README.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+export interface RunnerTestContext extends TestContext {
+  pdRunner: PodmanDesktopRunner;
+}
+```
+
+```extended-hook.ts
+import type { RunnerTestContext } from '../testContext/runner-test-context';
+import { afterEach } from 'vitest';
+import { takeScreenshotHook } from 'podman-desktop-tests-playwright';
```
Comment (benoitf) [inline tests/playwright/README.md]:
```suggestion
import { takeScreenshotHook } from '@podman-desktop/tests-playwright';
```

File: tests/playwright/README.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+  context.onTestFailed(async () => await takeScreenshotHook(context.pdRunner, context.task.name));
+});
+```
+
+### Global Setup file configuraiton
+
+Adding Global Setup/teardown module
+
+```global-setup.ts
+import { removeFolderIfExists } from 'podman-desktop-tests-playwright';
```
Comment (benoitf) [inline tests/playwright/README.md]:
```suggestion
import { removeFolderIfExists } from '@podman-desktop/tests-playwright';
```

File: tests/playwright/README.md [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+### Vitest Config
+
+Example of the vitest test configuration file.
+
+```
+const config = {
+  test: {
+    globals: true,
+    globalSetup: './path/to/globalSetup/global-setup.ts',
+    setupFiles: './path/to/hooks/extended-hooks.ts',
```
Comment (benoitf) [inline tests/playwright/README.md]: should it be in node_modules/@podman-desktop/tests-playwright ?

File: tests/playwright/README.md [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+## How to build locally
+
+1. Node 20 (ideal is to use `nvm`)
+2. Install local dependencies: `npm install`
+3. Build: `npm run build`
+4. Create Local package `npm run package`
+5. Connect to the repository where you want to write your tests
+
+## Usage
```
Comment (benoitf) [inline tests/playwright/README.md]: it's not clear in usage if it's when using in this repository or when wanting to embed/use in a 3rd party extension repository

File: tests/playwright/package.json [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    "build": "rimraf dist/ && tsc",
+    "package": "npm run build && npm link && npm pack",
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "publish:next": "yarn publish --registry=https://registry.npmjs.org/ --no-git-tag-version --new-version 0.0.1-\"$(date +%s)\""
+  },
+  "dependencies": {
+    "electron": "29.1.4",
+    "rimraf": "5.0.5",
+    "typescript": "^5.4.2",
+    "vitest": "^1.3.1"
```
Comment (benoitf) [inline tests/playwright/package.json]: should be dev dependencies

File: tests/playwright/tsconfig.json [hunk 8]
```diff
@@ -1,16 +1,14 @@
...
+    "moduleResolution": "node",
+    "rootDir": "./src",
+    "outDir": "./dist",
     "skipLibCheck": true,
-    "allowSyntheticDefaultImports": true,
-    "resolveJsonModule": true,
     "strict": true,
-    "isolatedModules": true,
+    "declaration": true,
+    "allowSyntheticDefaultImports": true,
```
Comment (benoitf) [inline tests/playwright/tsconfig.json]: not sure why we need these changes ?

File: tests/playwright/yarn.lock [hunk 9]
```diff
@@ -0,0 +1,1442 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (benoitf) [inline tests/playwright/yarn.lock]: this file should not be there

**Comment by odockal** [comment]
@benoitf I have included the review feedback. In a meantime I am still encountering problem with compiling the ts. files. when I run `yarn build`, so I got everything in `dist` to try it locally, I am encountering the issue:
```
$ rimraf dist/ && tsc
src/setupFiles/extended-hooks.ts:47:54 - error TS2339: Property 'task' does not exist on type 'RunnerTestContext'.

47   await takeScreenshotHook(context.pdRunner, context.task.name);
                                                        ~~~~


Found 1 error in src/setupFiles/extended-hooks.ts:47

error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
So I try to compile the file itself to see more concrete problems:
```
➜  playwright git:(pde2e-module-2) ✗ npx tsc src/testContext/runner-test-context.ts
node_modules/vite/dist/node/index.d.ts:5:41 - error TS2307: Cannot find module 'rollup/parseAst' or its corresponding type declarations.
  There are types at '/home/odockal/git/podman-desktop/tests/playwright/node_modules/rollup/dist/parseAst.d.ts', but this result could not be resolved under your current 'moduleResolution' setting. Consider updating to 'node16', 'nodenext', or 'bundler'.

5 export { parseAst, parseAstAsync } from 'rollup/parseAst';
                                          ~~~~~~~~~~~~~~~~~


Found 1 error in node_modules/vite/dist/node/index.d.ts:5
```

not sure if we need to have eventually dist folder with compiled sources.

**Review by benoitf — COMMENTED** [review]
@odockal I think you're missing the packaging side with a bundler like 'vite' tool

so in `build` script instead of tsc it should probably be `vite build` with a `vite.config.js` file being added

**Review by cbr7 — APPROVED** [review]

File: tests/playwright/package.json [hunk 10]
```diff
@@ -0,0 +1,29 @@
+{
+  "name": "@podman-desktop/tests-playwright",
+  "version": "1.8.0",
```
Comment (benoitf) [inline tests/playwright/package.json]: it needs to be 1.9.0-next

File: tests/playwright/src/runner/podman-desktop-runner.ts [hunk 11]
```diff
@@ -56,10 +56,21 @@ export class PodmanDesktopRunner {
     try {
       // start the app with given properties
       this._running = true;
+      console.log(`### STARTING PODMAN DESKTOP APP ###`);
```
Comment (benoitf) [inline tests/playwright/src/runner/podman-desktop-runner.ts]:
```suggestion
      console.log('### STARTING PODMAN DESKTOP APP ###');
```

File: tests/playwright/src/runner/podman-desktop-runner.ts [hunk 12]
```diff
@@ -56,10 +56,21 @@ export class PodmanDesktopRunner {
...
+      console.log(`Electron app launch options: `);
```
Comment (benoitf) [inline tests/playwright/src/runner/podman-desktop-runner.ts]:
```suggestion
      console.log('Electron app launch options: ');
```

File: tests/playwright/src/runner/podman-desktop-runner.ts (hunk 11 above)
Comment (benoitf) [reply tests/playwright/src/runner/podman-desktop-runner.ts]: I would avoid the usage of ###

File: tests/playwright/src/setupFiles/extended-hooks-utils.ts [hunk 13]
```diff
@@ -0,0 +1,43 @@
+/**********************************************************************
+ * Copyright (C) 2023-2024-2024 Red Hat, Inc.
```
Comment (benoitf) [inline tests/playwright/src/setupFiles/extended-hooks-utils.ts]:
```suggestion
 * Copyright (C) 2023-2024 Red Hat, Inc.
```

File: tests/playwright/tsconfig.eslint.json [hunk 14]
```diff
@@ -0,0 +1,4 @@
+{
+  "include": ["src/**/*.ts", "../types/**/*.d.ts"],
+  "exclude": ["node_modules"]
+}
```
Comment (benoitf) [inline tests/playwright/tsconfig.eslint.json]: why do we need a tsconfig.eslint.json file ?

File: tests/playwright/tsconfig.json [hunk 15]
```diff
@@ -19,6 +20,16 @@
...
+  "include": [
+    "src/playwright/globalSetup/*.ts",
+    "src/playwright/model/**/*.ts",
+    "src/playwright/runner/*.ts",
+    "src/playwright/setupFiles/setupFiles.ts",
+    "src/playwright/setupFiles/extended-hooks-utils.ts",
+    "src/playwright/testContext/*.ts",
+    "src/playwright/utility/*.ts",
+    "src/*.ts",
+    "../types/**/*.d.ts"
```
Comment (benoitf) [inline tests/playwright/tsconfig.json]: why this change ?
I mean also why no global wildcards

File: tests/playwright/vite.config.js [hunk 16]
```diff
@@ -0,0 +1,43 @@
+import { join } from 'path';
```
Comment (benoitf) [inline tests/playwright/vite.config.js:19]: should raise a flag on a missing copyright header

File: tests/playwright/vite.config.js [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+import { builtinModules } from 'module';
+import typescript from '@rollup/plugin-typescript';
```
Comment (benoitf) [inline tests/playwright/vite.config.js:21]: why ?

**Review by benoitf — CHANGES_REQUESTED** [review]
tests/playwright/yarn.lock should be removed

and 

website-argos/test-1.spec.ts

File: tests/playwright/tsconfig.json (hunk 15 above)
Comment (odockal) [reply tests/playwright/tsconfig.json]: See the comment above, the reason was to exclude specs from dist, but also to have them eslint check. I guess I can make it better, this was a solution without another tsconfig which could improve everything.

File: tests/playwright/tsconfig.eslint.json (hunk 14 above)
Comment (odockal) [reply tests/playwright/tsconfig.eslint.json]: So. The one tsconfig.json in `tests/playwright` is being used to pack/compile,etc. all the files in. But we do not want to have specs included in `dist`. Which can be done here. But then eslint complains about files being excluded. Adding specs into `eslintignore` is not right since we wanted them being checked. So the solution is to have two tsconfig files, one for packaging, the other for eslint to include all files just for check purposes.

File: tests/playwright/tsconfig.json (hunk 15 above)
Comment (benoitf) [reply tests/playwright/tsconfig.json]: I still don't understand why you don't want them in dist folder

you can keep default single tsconfig file

if you don't want to publish/include them, (I think it's the goal of your changes) it's not there that you need to do the changes, it's in the files section of the package.json

probably like here:
https://github.com/containers/podman-desktop/blob/main/packages/ui/package.json#L15-L16

File: tests/playwright/vite.config.js (hunk 17 above)
Comment (odockal) [reply tests/playwright/vite.config.js:21]: So it includes all typescript files in building the final `dist` with vite. Does not work without it.

Comment (benoitf) [reply tests/playwright/vite.config.js:21]: when you say 'all files' ? which one ?

b/c we're not using this plugin in other package

File: tests/playwright/tsconfig.eslint.json (hunk 14 above)
Comment (benoitf) [reply tests/playwright/tsconfig.eslint.json]: I think you can remove this one now

File: tests/playwright/tsconfig.json (hunk 15 above)
Comment (odockal) [reply tests/playwright/tsconfig.json]: This works! perfect then.

File: tests/playwright/vite.config.js (hunk 17 above)
Comment (odockal) [reply tests/playwright/vite.config.js:21]: All `.ts` files under tests/playwright? I am not really sure how you do that in other packages, nor what is right, this works. without it it does not work. I do not know how better describe this.

File: tests/playwright/tsconfig.json (hunk 15 above)
Comment (odockal) [reply tests/playwright/tsconfig.json]: Update: This does not work, without adding files explicitly in tsconfig.json `exclude` section, they will be added into `dist`. Adding them into glob in package.json, with an exclamation mark unfortunately does not work.
We do not have use for specs in other places. I have not found an use case where we could use them. Seems odd to have tests part of library. But these are not big issue, I admit. Problematic is `extended-hooks.ts` file which serves as extension to a vitest afterEach functionality. It must be kept in this repo and not be exported into `dist`. Have not found better way.

Comment (odockal) [reply tests/playwright/tsconfig.json]: Update 2: ok, so the files wont make it into npm package. That could work. testing.

Comment (benoitf) [reply tests/playwright/tsconfig.json]: I think here is the global rule: tsconfig should include everything that we can compile with the project

then in vite config or files of package.json we can filter what needs to be included in the produced exported file that others will include

Comment (odockal) [reply tests/playwright/tsconfig.json]: So. if the files are included in tsconfig, and omitted via package.json, they do not appear in the npm package (archive). Which is ok, I guess.

Comment (benoitf) [reply tests/playwright/tsconfig.json]: note also that vite is a bundler, so it produces like a single file containing everything that needs to be published

if you need to restrict things, it'll be there

**Comment by odockal** [comment]
> tests/playwright/yarn.lock should be removed
> 
> and
> 
> website-argos/test-1.spec.ts

Alright! yarn.lock should have been removed and the argos test I was not sure. both are gone now.

File: tests/playwright/README.md [hunk 18]
```diff
@@ -0,0 +1,144 @@
+# Podman Desktop Playwright Tests
+
+Testing Framework dedicated to a Podman Desktop and its extensions.
+
+## How to build locally from podman-desktop repo
+
+0. git clone podman-desktop
+1. Install Node 20 (ideal is to use `nvm`)
```
Comment (benoitf) [inline tests/playwright/README.md]:
```suggestion
1. Install Node.js 20 (ideal is to use `nvm`)
```

File: tests/playwright/README.md (hunk 6 above)
Comment (benoitf) [reply tests/playwright/README.md]: it has been flagged as resolved but looks like there was no reply

AFAIK the README will be published on npmjs so I don't know why it says to clone podman-desktop repository

it should start at step 7

File: tests/playwright/package.json [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+  "scripts": {
+    "build": "rimraf dist/ && vite build",
+    "package": "npm run build && npm link && npm pack",
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "publish:next": "yarn publish --registry=https://registry.npmjs.org/ --no-git-tag-version --new-version 0.0.1-\"$(date +%s)\""
+  },
+  "dependencies": {
+    "electron": "29.1.4",
+    "rimraf": "5.0.5",
+    "typescript": "^5.4.2"
```
Comment (benoitf) [inline tests/playwright/package.json]: AFAIK these 2 are dev dependencies

Comment (benoitf) [reply tests/playwright/package.json]: I don't know then for tests if we should move also electron to a devDependency or if @playwright/test should be moved to dependency

probably all of them should be devDependencies and a peerDependencies should be added on electron and @playwright/test

File: tests/playwright/src/index.ts [hunk 20]
```diff
@@ -0,0 +1,60 @@
+/**********************************************************************
+ * Copyright (C) 2023-2024 Red Hat, Inc.
```
Comment (benoitf) [inline tests/playwright/src/index.ts]:
```suggestion
 * Copyright (C) 2024 Red Hat, Inc.
```

it's a new file, was not there in 2023

File: tests/playwright/src/runner/podman-desktop-runner.ts [hunk 21]
```diff
@@ -1,5 +1,5 @@
 /**********************************************************************
- * Copyright (C) 2023-2024 Red Hat, Inc.
+ * Copyright (C) 2023 Red Hat, Inc.
```
Comment (benoitf) [inline tests/playwright/src/runner/podman-desktop-runner.ts]:
```suggestion
 * Copyright (C) 2023-2024 Red Hat, Inc.
```

you're making changes to the file so 2023-2024 is correct

File: tests/playwright/src/runner/podman-desktop-runner.ts [hunk 22]
```diff
@@ -56,10 +56,21 @@ export class PodmanDesktopRunner {
     try {
       // start the app with given properties
       this._running = true;
+      console.log('### STARTING PODMAN DESKTOP APP ###');
```
Comment (benoitf) [inline tests/playwright/src/runner/podman-desktop-runner.ts]:
```suggestion
      console.log('Starting Podman Desktop App');
```

File: tests/playwright/vite.config.js (hunk 17 above)
Comment (benoitf) [reply tests/playwright/vite.config.js:21]: ok let's postpone this

File: tests/playwright/package.json (hunk 19 above)
Comment (odockal) [reply tests/playwright/package.json]: will try. thx

File: tests/playwright/README.md (hunk 6 above)
Comment (benoitf) [reply tests/playwright/README.md]: is there a reason why it's marked as resolved ?

**Review by benoitf — APPROVED** [review]
I think the README is not accurate if it's gonna be published as people should not have to clone podman-desktop

so probably need to start at step 7

**Comment by odockal** [comment]
> I think the README is not accurate if it's gonna be published as people should not have to clone podman-desktop
> 
> so probably need to start at step 7

I have update the README. So the first part is how to add contribution to the library, how to developer locally. Using it as an npm package if then the rest of the guide. Hope it is better readable now.

**Comment by benoitf** [comment]
@odockal probably you should have CONTRIBUTING.md for contributions and README for the consumers of the library

( can be follow-ups)

**Comment by odockal** [comment]
> @odockal probably you should have CONTRIBUTING.md for contributions and README for the consumers of the library
> 
> ( can be follow-ups)

Ok, here is a follow up issue #6466. Thanks for review!

File: tests/playwright/README.md [hunk 23]
```diff
@@ -0,0 +1,146 @@
+# Podman Desktop Playwright Tests
+
+Testing Framework dedicated to a Podman Desktop and its extensions.
+
+## How to developer using locally built @podman-desktop/tests-playwright
```
Comment (benoitf) [inline tests/playwright/README.md]:
```suggestion
## How to develop using locally built @podman-desktop/tests-playwright
```
